hypothesis-python/src/hypothesis/strategies/_internal/datetime.py

```diff
@@ -39,6 +39,32 @@ def is_pytz_timezone(tz):
     return module == "pytz" or module.startswith("pytz.")
 
 
+def replace_tzinfo(value, timezone):
+    if is_pytz_timezone(timezone):
+        # Can't just construct; see http://pytz.sourceforge.net
+        # After dropping Python 3.5 support, just use `result.fold`
+        return timezone.localize(value, is_dst=not getattr(value, "fold", 0))
```

Comment:
I don't think anything needs to change here, but `is_dst` is not *exactly* the inverse of `fold`. `fold` basically says whether in ambiguous or gap times you should use the offset that applies *before* the gap (`fold=0`) or the offset that applies *after* the gap (`fold=1`). `is_dst` says whether or not you should choose the side that is "DST" or "STD" (not sure what they do in STD->STD or DST->DST transitions).

I have some diagrams that may or may not be comprehensible. This is an example of what offsets apply when using PEP 495's conception of `fold`:

![fold_gap_scheme_2](https://user-images.githubusercontent.com/1377457/79125272-e8448a00-7d6b-11ea-8d45-f5b955f66291.png)

And here is an example of what offsets apply when you use "fold" to mean the inverse of `is_dst`:

![fold_gap_scheme_1](https://user-images.githubusercontent.com/1377457/79125263-e37fd600-7d6b-11ea-884a-c5ce30ed2218.png)

The difference is basically that it inverts what offsets are returned during imaginary times (note that the right half of the diagram is the same in both schemes). This doesn't really matter, though, particularly when you are generating the value *randomly* (I don't even think `fold` will be set on the result of this `localize` call). I think the only effect is that result shrinking will tend to favor the DST offset rather than the offset that applied before the transition, but that's a pretty arbitrary choice anyway.

Reply:
Oh no.

`is_dst=not val.fold` works *except* during DST-to-STD transitions involving an imaginary time, which happens due to the negative offset in e.g. `Europe/Dublin`.

...I've just documented that this can happen and that we recommend a PEP-495-compliant library like `dateutil`, but other suggestions are welcome.